src/MacroExpander.js

```diff
@@ -17,114 +17,194 @@ class MacroExpander {
+                        // to replace placeholder with the indicated argument.
+                        // TODO: use spread once we move to ES2015
+                        expansion.splice.apply(
+                            expansion,
+                            [i, 2].concat(args[tok.text - 1]));
                     } else {
-                        args[i] = [startOfArg];
+                        throw new ParseError(
+                            "Not a valid argument number",
+                            tok);
                     }
                 }
-                // paste arguments in place of the placeholders
-                expansion = expansion.slice(); // make a shallow copy
-                for (i = expansion.length - 1; i >= 0; --i) {
-                    tok = expansion[i];
-                    if (tok.text === "#") {
-                        if (i === 0) {
-                            throw new ParseError(
-                                "Incomplete placeholder at end of macro body",
-                                tok);
-                        }
-                        tok = expansion[--i]; // next token on stack
-                        if (tok.text === "#") { // ## → #
-                            expansion.splice(i + 1, 1); // drop first #
-                        } else if (/^[1-9]$/.test(tok.text)) {
-                            // expansion.splice(i, 2, arg[0], arg[1], …)
-                            // to replace placeholder with the indicated argument.
-                            // TODO: use spread once we move to ES2015
-                            expansion.splice.apply(
-                                expansion,
-                                [i, 2].concat(args[tok.text - 1]));
-                        } else {
-                            throw new ParseError(
-                                "Not a valid argument number",
-                                tok);
-                        }
-                    }
+            }
+        }
+        // Concatenate expansion onto top of stack.
+        this.stack.push.apply(this.stack, expansion);
+        return expansion;
+    }
+
+    /**
+     * Expand the next token once (ignoring initial spaces like `get`),
+     * without removing anything from the stack, and return the top token
+     * on the stack.  Similar in behavior to TeX's `\expandafter\futurelet`.
+     */
+    expandAfterFuture() {
+        this.expandOnce();
+        return this.future();
+    }
+
+    /**
+     * Recursively expand first token, then return first non-expandable token.
+     */
+    expandNextToken() {
+        for (;;) {
+            const expanded = this.expandOnce();
+            // expandOnce returns Token if and only if it's fully expanded.
+            if (expanded instanceof Token) {
+                // \relax stops the expansion, but shouldn't get returned (a
+                // null return value couldn't get implemented as a function).
```

Comment:
I'm not sure what `a null return value couldn't get implemented as a function` means.  Could you give an example?

Reply:
Functions (as defined by `defineFunction`) return a `ParseNode` or an object that gets converted into a `ParseNode` by the parser.  It's not possible for a function to return `null` that would just disappear in the final parse tree.  Hence it would be impossible to implement a `\relax` that expands to nothing (which is different to how `{}` expands).

Reply:
So if we return `null` the parser would think that it's an object an convert it to a `ParseNode`.  I wonder when would we ever want to write a function that returns a `null`?

Reply:
Long-term, I vaguely hope we can transition functions over to macros, as that's what TeX really does.  So maybe not too critical... though certainly could be done by tweaking group parsing.